File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Queue;

class Solution {
    public int solution(int n, int[][] edge) {
        int answer = 0; 
        /**
        *   1번 노드에 대해 각 노드까지의 최단거리를 다익스트라를 활용해 구할 수 있다.
        */
        
        int[] cost = new int[n + 1];
        List<Integer>[] map = new ArrayList[n + 1];
        
        Arrays.fill(cost, Integer.MAX_VALUE);
        for (int i = 1; i <= n; i++) {
            map[i] = new ArrayList<>();
        }
        
        for (int[] e: edge) {
            map[e[0]].add(e[1]);
            map[e[1]].add(e[0]);
        }
        
        int max = bfs(map, cost);
        for (int i = 2; i <= n; i++) {
            if (max == cost[i]) {
                answer++;
            }
        }
        
        return answer;
    }
    
    int bfs(List<Integer>[] map, int[] cost) {
        Queue<Integer> q = new ArrayDeque<>();
        cost[1] = 0;
        q.add(1);
        
        int max = 0;
        while (!q.isEmpty()) {
            int currValue = q.poll();
            
            for (int nextValue: map[currValue]) {
                if (cost[nextValue] == Integer.MAX_VALUE) {
                    int nextCost = cost[currValue] + 1;
                    cost[nextValue] = nextCost;
                    q.add(nextValue);
                    
                    max = Math.max(nextCost, max);
                }
            }
        }
        
        return max;
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(int n, vector<vector<int>> edge) {
    int answer = 0;
    /**
    *   간선의 비용이 모두 1이므로, BFS를 통해 탐색 가능하다.
    *   큐에서 1에서부터 같은 비용을 가지는 노드들을 세트로 꺼내 BFS 탐색을 진행한다.
    */
    
    // (i, j)를 연결하는 간선을 표시하는 인접 배열
    vector<vector<bool>> map (n + 1, vector<bool> (n + 1));
    vector<bool> visited (n + 1);
    for (auto& e: edge) {
        int x = e[0], y = e[1];
        
        map[x][y] = true;
        map[y][x] = true;
    }

    queue<int> q;
    q.push(1);
    visited[1] = true;
    
    int count = 1;
    while (!q.empty()) {
        count = q.size();
        answer = count;
        
        // 1로부터의 특정 거리인 노드들만 한 그룹으로 큐에서 꺼낸다.
        while (count-- > 0) {
            int curr = q.front();
            q.pop();
            
            for (int i = 1; i <= n; i++) {
                if (!visited[i] && map[curr][i]) {
                    visited[i] = true;
                    q.push(i);
                }
            }
        }
    }
    
    return answer;
}